<?xml version="1.0" ?>
<launch>
  <!-- LAUNCH INTERFACE -->

  <!-- The robot name is required for gazebo to spawn a model. -->
  <arg name="robotName" default="lwr"/>
  <arg name="robotOrigin.x" default="0"/>
  <arg name="robotOrigin.y" default="0"/>
  <arg name="robotOrigin.z" default="0"/>
  <arg name="robotOrigin.ax" default="0"/>
  <arg name="robotOrigin.ay" default="0"/>
  <arg name="robotOrigin.az" default="0"/>  
  <arg name="leftRobotName" default="lwr2"/>
  <arg name="leftRobotOrigin.x" default="0.066"/>
  <arg name="leftRobotOrigin.y" default="0.9"/>
  <arg name="leftRobotOrigin.z" default="0"/>
  <arg name="leftRobotOrigin.ax" default="0"/>
  <arg name="leftRobotOrigin.ay" default="0"/>
  <arg name="leftRobotOrigin.az" default="0"/>

  <arg name="trocarOffset.x" default="-0.5"/>
  <arg name="trocarOffset.y" default="0.0"/>
  <arg name="trocarOffset.z" default="0.1"/>
  <arg name="trocarOffset2.x" default="-0.5"/>
  <arg name="trocarOffset2.y" default="0.15"/>
  <arg name="trocarOffset2.z" default="0.1"/>
  <arg name="trocarOffset3.x" default="-0.5"/>
  <arg name="trocarOffset3.y" default="-0.15"/>
  <arg name="trocarOffset3.z" default="0.1"/>

  <!-- The recommended way of specifying the URDF file is to pass it in
       directly. -->
  <arg name="robotPath" default="$(find robotic_experiments)/models"/>
  <arg name="lwrRobotUrdfFile" default="$(arg robotPath)/lwr_with_tool.urdf.xacro" />

  <arg name="controllers" default="joint_controllers"/>
  <arg name="t1_limits" default="false"/>

  <arg name="rviz_config" default="$(find lwr_launch)/launch/rviz/rviz_config.rviz"/>
  <arg name="rviz_bringup" default="true"/> <!--If false, do not launch rviz-->

  <arg name="GAZEBO_GUI" default="false"/>


  <arg name="hw_interface_file" default="$(find single_lwr_robot)/config/hw_interface.yaml"/>
  <arg name="controller_config_file" default="$(find single_lwr_robot)/config/controllers.yaml"/>
  <arg name="gazebo_world_file" default="$(find robotic_experiments)/world/test_world.world"/>
  <arg name="t1_limits_file" default="$(find single_lwr_robot)/config/t1_joint_limits.yaml"/>


  <!-- Load updated joint limits (override information from single_lwr_moveit) to respect T1 mode limits -->
  <group if="$(arg t1_limits)" ns="robot_description_planning">
    <rosparam command="load" file="$(arg t1_limits_file)"/>
  </group>

  <!-- load robotHW configurations to rosparam server -->
  <rosparam command="load" file="$(arg hw_interface_file)"/>

  <!-- load all controller configurations to rosparam server -->
  <rosparam command="load" file="$(arg controller_config_file)"/>

  <arg name="lwr2_hw_interface_file" default="$(find robotic_experiments)/config/lwr2_hw_interface.yaml"/>
  <arg name="lwr2_controller_config_file" default="$(find robotic_experiments)/config/lwr2_controllers.yaml"/>
  <rosparam command="load" file="$(arg lwr2_hw_interface_file)"/>
  <rosparam command="load" file="$(arg lwr2_controller_config_file)"/>


    <!-- <node name="worldRightTransform" pkg="foot_control_xy" type="worldActorTransform" args="foot -x 2.0 -y 0.0 -z 0 -qx 0.0 -qy 0.0 -qz 0.0 -qw 1.0"/> -->
  <!-- ROBOT -->
  <group ns="lwr">
    <param name="robot_description" command="$(find xacro)/xacro --check-order --inorder $(arg lwrRobotUrdfFile) robotName:=$(arg robotName) origin.x:=$(arg robotOrigin.x) origin.y:=$(arg robotOrigin.y) origin.z:=$(arg robotOrigin.z) origin.ax:=$(arg robotOrigin.ax) origin.ay:=$(arg robotOrigin.ay) origin.az:=$(arg robotOrigin.az) addCamera:=1 addFrame:=1"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="controller_spawner_js" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="--namespace=/lwr kuka_joint_state_controller"/>
    <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="--namespace=/lwr $(arg controllers)"/>
    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param /lwr/robot_description -urdf -model $(arg robotName)" respawn="false" output="screen" />
<!--     <node name="publish_trocar_frame" pkg="surgical_simulator" type="publish_trocar_frame" args="trocar -x $(arg trocarOffset.x) -y $(arg trocarOffset.y) -z $(arg trocarOffset.z)"/>
    <node name="publish_trocar_frame2" pkg="surgical_simulator" type="publish_trocar_frame" args="trocar2 -x $(arg trocarOffset2.x) -y $(arg trocarOffset2.y) -z $(arg trocarOffset2.z)"/>
    <node name="publish_trocar_frame3" pkg="surgical_simulator" type="publish_trocar_frame" args="trocar3 -x $(arg trocarOffset3.x) -y $(arg trocarOffset3.y) -z $(arg trocarOffset3.z)"/> -->
  </group>


  <group ns="lwr2">
    <param name="robot_description" command="$(find xacro)/xacro --check-order --inorder $(arg lwrRobotUrdfFile) robotName:=$(arg leftRobotName) origin.x:=$(arg leftRobotOrigin.x) origin.y:=$(arg leftRobotOrigin.y) origin.z:=$(arg leftRobotOrigin.z) origin.ax:=$(arg leftRobotOrigin.ax) origin.ay:=$(arg leftRobotOrigin.ay) origin.az:=$(arg leftRobotOrigin.az) addCamera:=1"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="controller_spawner_js" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="kuka_joint_state_controller"/>
    <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="--namespace=/lwr2 $(arg controllers)"/>
    <node name="left_spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param /lwr2/robot_description -urdf -model $(arg leftRobotName)" respawn="false" output="screen" />
  </group>

  <!-- GAZEBO -->
  <!-- enable/disable gui at will, the rviz listens to the simulation -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(arg gazebo_world_file)"/>
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="$(arg GAZEBO_GUI)"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
  </include>

  <!-- <node pkg="tf" type="static_transform_publisher" name="right_to_left_transform" args="$(arg leftRobotOrigin.x) $(arg leftRobotOrigin.y) $(arg leftRobotOrigin.z) 0 0 0 /world /left_robot_basis 200"/>  -->


<!--   <group ns="pancreas">
    <param name="robot_description" command="$(find xacro)/xacro.py $(arg pancreasUrdfFile)"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="pancreas_spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-x 3 -y 3 -z 3 -param /pancreas/robot_description -urdf -model pancreas" respawn="false" output="screen" />
  </group> -->

  <!-- LAUNCH RVIZ -->

  <node name="lwr_rviz" pkg="rviz" type="rviz" respawn="false" args="-d $(arg rviz_config)"
        output="screen"
        if="$(arg rviz_bringup)"/>

    <!-- <node name="righTrocarPublish" pkg="surgical_simulator" type="publish_trocar_frame" args="right_trocar -x 0 -y 0 -z 0"/> -->

</launch>
